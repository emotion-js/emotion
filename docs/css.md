---
title: "CSS"
---

The primary way to style things in emotion is with `css`, it accepts styles as a template literal, object, or array of objects and returns a class name.

`css` can be used as a [tagged template literal](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals).

```jsx live
import { css } from 'emotion'

const color = 'darkgreen'

render(
  <div
    className={css`
      background-color: hotpink;
      &:hover {
        color: ${color};
      }
    `}
  >
    This has a hotpink background.
  </div>
)
```

`css` also accepts styles as an object, for more usage with objects, go [here](/docs/object-styles.md).

```jsx live
import { css } from 'emotion'

render(
  <div
    className={css({
      backgroundColor: 'hotpink',
      '&:hover': {
        color: 'lightgreen'
      }
    })}
  >
    This has a hotpink background.
  </div>
)
```

## CSS Prop

> Note:

> [The css prop requires `babel-plugin-emotion`](/packages/babel-plugin-emotion).


With `babel-plugin-emotion`, the css prop can be used, it accepts styles like `css` and adds it to the className of the element it's on. This happens at compile time by converting the css prop to a css call and prepending it to the className of the element. It will only work if you use it as an actual JSX attribute, if it's in an object that's spread onto the element it won't work.

```jsx live
function SomeComponent(props) {
  // Create styles as if you're calling css and the class will be applied to the component
  return (
    <div
      css={`
    color: blue;
    font-size: ${props.fontSize}px;

    &:hover {
      color: green;
    }

    & .some-class {
      font-size: 20px;
    }
  `}
    >
      This will be blue until hovered.
      <div className="some-class">This font size will be 20px</div>
      <div css={{ color: 'hotpink' }}>This is hotpink</div>
    </div>
  )
}
render(<SomeComponent fontSize={15} />)
```

> Note:

> The css prop is not compatible with `babel-plugin-transform-react-inline-elements`. If you include it in your `.babelrc` no transformation will take place and your styles won't be applied.