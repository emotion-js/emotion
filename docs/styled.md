---
title: "Styled Components"
---

`styled` is a way to create React or Preact components that have styles attached to them. It's available from [react-emotion](/packages/react-emotion) and [preact-emotion](/packages/preact-emotion). `styled` was heavily inspired by [styled-components](https://www.styled-components.com/) and [glamorous](https://glamorous.rocks/)

### Styling elements and components

`styled` is very similar to `css` except you call it with an html tag or React/Preact component and then call that with a template literal for string styles or a regular function call for object styles.

```jsx live
import styled from 'react-emotion'

const Button = styled('button')`
  color: turquoise;
`

render(<Button>This my button component.</Button>)
```

### Changing based on props

Any interpolations or arguments that are functions in `styled` are called with `props`, this allows you to change the styles of a component based on the props.

```jsx live
import styled from 'react-emotion'

const Button = styled('button')`
  color: ${props =>
    props.primary ? 'hotpink' : 'turquoise'};
`

const Container = styled('div')(props => ({
  display: 'flex',
  flexDirection: props.column && 'column'
}))

render(
  <Container column>
    <Button>
      This is a regular button.
    </Button>
    <Button primary>
      This is a primary button.
    </Button>
  </Container>
)
```

### Styling any component

`styled` can style any component as long as it accepts a `className` prop.

```jsx live
import styled from 'react-emotion'
const Basic = ({ className }) => (
  <div className={className}>Some text</div>
)

const Fancy = styled(Basic)`
  color: hotpink;
`

render(<Fancy />)
```


### Change the rendered tag using `withComponent`

Sometimes you want to create some styles with one component but then use those styles again with another component, the `withComponent` method can be used for this. This was inspired by [styled-components' `withComponent`](https://www.styled-components.com/docs/api#withcomponent).

```jsx live
// Create a section element
const Section = styled('section')`
  background: #333;
`
// Create an aside element with the same styles as Section
const Aside = Section.withComponent('aside')
render(
  <div>
    <Section>This is a section</Section>
    <Aside>This is an an aside</Aside>
  </div>
)
```

### Targeting another emotion component

Similar to [styled-components](https://www.styled-components.com/docs/faqs#can-i-refer-to-other-components), emotion allows for emotion components to be targeted like regular CSS selectors when using [babel-plugin-emotion](/packages/babel-plugin-emotion.md).

```jsx live
const Child = styled('div')`
  color: red;
`

const Parent = styled('div')`
  ${Child} {
    color: green;
  }
`
render(
  <div>
    <Parent>
      <Child>green</Child>
    </Parent>
    <Child>red</Child>
  </div>
)
```

Component selectors can also be used with object styles.

```jsx live
const Child = styled('div')({
  color: 'red'
})

const Parent = styled('div')({
  [Child]: {
    color: 'green'
  }
})

render(
  <div>
    <Parent>
      <Child>green</Child>
    </Parent>
    <Child>red</Child>
  </div>
)
```

### Pass refs down using `innerRef`

Sometimes you need to get a [ref](https://reactjs.org/docs/refs-and-the-dom.html) but passing `ref` to a styled component will return a ref to the styled component, not the component that it renders which is generally the one you want. You can pass `innerRef` instead of `ref` to get the ref of the component that styled renders.

```jsx live
const Input = styled('input')`
  color: hotpink;
`

const Button = styled('button')`
  color: green;
`

function TextInput(props) {
  let textInput = null
  function handleClick() {
    textInput.focus()
  }
  return (
    <div>
      <Input
        innerRef={input => {
          textInput = input
        }}
      />
      <Button onClick={handleClick}>
        Focus the text input
      </Button>
    </div>
  )
}
render(<TextInput />)
```

### Element Shorthand

> Note:

> `babel-plugin-emotion` is required for the element shorthand

Instead of using the function call syntax(`styled('div')`), you can use create components by using a property, where the property refers to an HTML tag(`styled.div`).

```jsx live
const DivWithoutShorthand = styled('div')`
  color: green;
`

const DivWithShorthand = styled.div`
  color: hotpink;
`

render(
  <DivWithoutShorthand>
    This is green. <DivWithShorthand>This is hotpink.</DivWithShorthand>
  </DivWithoutShorthand>
)
```

### Object styles

```jsx live
import styled from 'react-emotion'

const H1 = styled.h1(
  {
    fontSize: 20
  },
  props => ({ color: props.color })
)

render(
  <H1 color="lightgreen">
    This is lightgreen.
  </H1>
)
```

This API was inspired by [glamorous](https://github.com/paypal/glamorous).

### withConfig is not a function error

This error is caused by using the shorthand syntax for styled such as `styled.div` without the Babel plugin. To fix this, [install `babel-plugin-emotion`](/docs/babel.md)
