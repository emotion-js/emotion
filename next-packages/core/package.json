{
  "name": "@emotion/core",
  "version": "0.10.4",
  "main": "dist/index.cjs.js",
  "module": "dist/index.es.js",
  "browser": {
    "./dist/index.cjs.js": "./dist/index.browser.cjs.js",
    "./dist/index.es.js": "./dist/index.browser.es.js"
  },
  "files": [
    "src",
    "dist"
  ],
  "author": "mitchellhamilton <mitchell@mitchellhamilton.me>",
  "license": "MIT",
  "scripts": {
    "build": "npm-run-all clean rollup",
    "clean": "rimraf dist",
    "rollup": "rollup -c ../../rollup.config.js"
  },
  "dependencies": {
    "@emotion/cache": "^0.8.3",
    "@emotion/css": "^0.9.3",
    "@emotion/serialize": "^0.8.3",
    "@emotion/sheet": "^0.6.6",
    "@emotion/utils": "^0.7.1"
  },
  "peerDependencies": {
    "react": ">=16.3.0"
  },
  "devDependencies": {
    "html-tag-names": "^1.1.2",
    "svg-tag-names": "^1.1.1"
  },
  "repository": "https://github.com/emotion/next/tree/master/packages/core",
  "publishConfig": {
    "access": "public"
  },
  "umd:main": "./dist/index.min.js"
}
