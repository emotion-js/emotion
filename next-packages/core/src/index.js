// @flow
export { withCSSContext, Provider } from './context'
export { jsx } from './jsx'
export { Global } from './global'
export { keyframes } from './keyframes'
export { default as css } from '@emotion/css'
