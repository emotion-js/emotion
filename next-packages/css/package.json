{
  "name": "@emotion/css",
  "version": "0.9.3",
  "description": "a function to serialize css and object styless",
  "main": "dist/index.cjs.js",
  "module": "dist/index.es.js",
  "license": "MIT",
  "repository": "https://github.com/emotion-js/next/tree/master/packages/css",
  "dependencies": {
    "@emotion/hash": "^0.6.3",
    "@emotion/serialize": "^0.8.3",
    "@emotion/utils": "^0.7.1"
  },
  "publishConfig": {
    "access": "public"
  },
  "files": [
    "src",
    "dist"
  ],
  "umd:main": "./dist/index.min.js"
}
