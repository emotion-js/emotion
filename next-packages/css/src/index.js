// @flow
import type { Interpolation, ScopedInsertableStyles } from '@emotion/utils'
import { serializeStyles } from '@emotion/serialize'

let fakeRegisteredCache = {}

function css(...args: Array<Interpolation>): ScopedInsertableStyles {
  return serializeStyles(fakeRegisteredCache, args)
}

export default css
