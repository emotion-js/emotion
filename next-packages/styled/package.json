{
  "name": "@emotion/styled",
  "version": "0.9.3",
  "description": "styled API for emotion",
  "main": "dist/index.cjs.js",
  "module": "dist/index.es.js",
  "license": "MIT",
  "repository": "https://github.com/emotion-js/next/tree/master/packages/styled",
  "dependencies": {
    "@emotion/styled-base": "^0.9.3"
  },
  "publishConfig": {
    "access": "public"
  },
  "files": [
    "src",
    "dist"
  ],
  "umd:main": "./dist/index.min.js"
}
