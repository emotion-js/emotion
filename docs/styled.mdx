---
title: 'Styled Components'
---

`styled` is a way to create React components that have styles attached to them. It's available from [@emotion/styled](/packages/@emotion/styled). `styled` was heavily inspired by [styled-components](https://www.styled-components.com/) and [glamorous](https://glamorous.rocks/)

### Styling elements and components

`styled` is very similar to `css` except you call it with an html tag or React component and then call that with a template literal for string styles or a regular function call for object styles.

```jsx
// @live
import styled from '@emotion/styled'

const Button = styled.button`
  color: turquoise;
`

render(<Button>This my button component.</Button>)
```

### Changing based on props

Any interpolations or arguments that are functions in `styled` are called with `props`, this allows you to change the styles of a component based on the props.

```jsx
// @live
import styled from '@emotion/styled'

const Button = styled.button`
  color: ${props =>
    props.primary ? 'hotpink' : 'turquoise'};
`

const Container = styled.div(props => ({
  display: 'flex',
  flexDirection: props.column && 'column'
}))

render(
  <Container column>
    <Button>This is a regular button.</Button>
    <Button primary>This is a primary button.</Button>
  </Container>
)
```

### Styling any component

`styled` can style any component as long as it accepts a `className` prop.

```jsx
// @live
import styled from '@emotion/styled'
const Basic = ({ className }) => (
  <div className={className}>Some text</div>
)

const Fancy = styled(Basic)`
  color: hotpink;
`

render(<Fancy />)
```

### Change the rendered tag using `withComponent`

Sometimes you want to create some styles with one component but then use those styles again with another component, the `withComponent` method can be used for this. This was inspired by [styled-components' `withComponent`](https://www.styled-components.com/docs/api#withcomponent).

```jsx
// @live
import styled from '@emotion/styled'

const Section = styled.section`
  background: #333;
`
// this component has the same styles as Section but it renders an aside
const Aside = Section.withComponent('aside')
render(
  <div>
    <Section>This is a section</Section>
    <Aside>This is an an aside</Aside>
  </div>
)
```

### Targeting another emotion component

Similar to [styled-components](https://www.styled-components.com/docs/faqs#can-i-refer-to-other-components), emotion allows for emotion components to be targeted like regular CSS selectors when using [babel-plugin-emotion](/packages/babel-plugin-emotion.md).

```jsx
// @live
import styled from '@emotion/styled'

const Child = styled.div`
  color: red;
`

const Parent = styled.div`
  ${Child} {
    color: green;
  }
`
render(
  <div>
    <Parent>
      <Child>green</Child>
    </Parent>
    <Child>red</Child>
  </div>
)
```

Component selectors can also be used with object styles.

```jsx
// @live
import styled from '@emotion/styled'

const Child = styled.div({
  color: 'red'
})

const Parent = styled.div({
  [Child]: {
    color: 'green'
  }
})

render(
  <div>
    <Parent>
      <Child>green</Child>
    </Parent>
    <Child>red</Child>
  </div>
)
```

### Object styles

```jsx
// @live
import styled from '@emotion/styled'

const H1 = styled.h1(
  {
    fontSize: 20
  },
  props => ({ color: props.color })
)

render(<H1 color="lightgreen">This is lightgreen.</H1>)
```

This API was inspired by [glamorous](https://github.com/paypal/glamorous). ❤️

### Customizing prop forwarding

By default, Emotion passes all props to custom components and only props that are valid html attributes for string tags. You can customise this by passing a custom `shouldForwardProp` function. You can also use `@emotion/is-prop-valid` (which is used by emotion internally) to filter out props that are not valid as html attributes.

```jsx
// @live
import isPropValid from '@emotion/is-prop-valid'
import styled from '@emotion/styled'

const H1 = styled('h1', {
  shouldForwardProp: prop =>
    isPropValid(prop) && prop !== 'color'
})(props => ({
  color: 'hotpink'
}))

render(<H1 color="lightgreen">This is lightgreen.</H1>)
```

### Composing dynamic styles

You can create dynamic styles that are based on props and use them in styles.

```jsx
// @live
import styled from '@emotion/styled'
import { css } from '@emotion/core'

const dynamicStyle = props =>
  css`
    color: ${props.color};
  `

const Container = styled.div`
  ${dynamicStyle};
`
render(
  <Container color="lightgreen">
    This is lightgreen.
  </Container>
)
```

### `as` prop

To use styles from a styled component but change the element that's rendered, you can use the `as` prop.

```jsx
// @live
import styled from '@emotion/styled'

const Button = styled.button`
  color: hotpink;
`

render(
  <Button
    as="a"
    href="https://github.com/emotion-js/emotion"
  >
    Emotion on GitHub
  </Button>
)
```

This API was inspired by [styled-components](https://www.styled-components.com). ❤️

The `as` prop is only used by styled when it's not forwarded to the underlying element. By default, this means that the `as` prop is used for html tags and forwarded for components. To change this, you can pass a custom [`shouldForwardProp`](#customizing-prop-forwarding) which returns `true` for `'as'` to forward it or returns `false` for `'as'` to use it and not forward it.

### Nesting components

We can nest selectors using `&`:

```jsx
// @live

import styled from '@emotion/styled'

const Example = styled('span')`
  color: lightgreen;
  & > a {
    color: hotpink;
  }
`

render(
  <Example>
    This is <a>nested</a>.
  </Example>
)
```
