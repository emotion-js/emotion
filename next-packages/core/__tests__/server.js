/** @jsx jsx
 * @jest-environment node
 */
// @flow
import 'test-utils/dev-mode'
import * as React from 'react'
import cases from 'jest-in-case'
import { jsx, Global, keyframes, CSSContext } from '@emotion/core'
import styled from '@emotion/styled'
import css from '@emotion/css'
import createCache from '@emotion/cache'
import { renderToString } from 'react-dom/server'
import HTMLSerializer from 'jest-serializer-html'

expect.addSnapshotSerializer(HTMLSerializer)

cases(
  'ssr',
  opts => {
    expect(renderToString(opts.render())).toMatchSnapshot()
  },
  {
    basic: {
      render: () => <div css={{ color: 'hotpink' }}>some hotpink text</div>
    },
    global: {
      render: () => (
        <Global
          styles={{
            html: {
              backgroundColor: 'hotpink'
            }
          }}
        />
      )
    },
    keyframes: {
      render: () => {
        const animation = keyframes(
          css`
            from {
              color: green;
            }
            to {
              color: blue;
            }
          `
        )

        return (
          <React.Fragment>
            <div
              css={[{ animation: `1s ${animation.name}` }, animation.styles]}
            />
          </React.Fragment>
        )
      }
    },
    'only render a style once with the css prop': {
      render: () => {
        return (
          <div css={{ color: 'green' }}>
            <div css={{ color: 'hotpink' }} />
            <div css={{ color: 'hotpink' }} />
          </div>
        )
      }
    },
    'only render a style once with styled': {
      render: () => {
        const SomeComponent = styled.div`
          color: hotpink;
        `
        return (
          <div css={{ color: 'green' }}>
            <SomeComponent />
            <SomeComponent />
          </div>
        )
      }
    },
    'works with nonces': {
      render: () => {
        const SomeComponent = styled.div`
          color: hotpink;
        `
        return (
          <CSSContext.Provider value={createCache({ nonce: 'some-nonce' })}>
            <SomeComponent />
            <div css={{ color: 'hotpink' }} />
            <Global
              styles={{
                html: {
                  margin: 0,
                  padding: 0,
                  fontFamily: 'sans-serif'
                }
              }}
            />
          </CSSContext.Provider>
        )
      }
    }
  }
)
