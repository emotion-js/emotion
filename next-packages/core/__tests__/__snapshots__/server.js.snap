// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ssr basic 1`] = `

<style data-emotion-css="1lrxbo5">
  .css-1lrxbo5{color:hotpink;}
</style>
<div class="css-1lrxbo5">
  some hotpink text
</div>

`;

exports[`ssr global 1`] = `

<style data-emotion-css="6l7sk3">
  html{background-color:hotpink;}
</style>

`;

exports[`ssr keyframes 1`] = `

<style data-emotion-css="1s775nx">
  .css-1s775nx{-webkit-animation:1s animation-1kyaq4y;animation:1s animation-1kyaq4y;}@-webkit-keyframes animation-1kyaq4y{from{color:green;}to{color:blue;}}@keyframes animation-1kyaq4y{from{color:green;}to{color:blue;}}
</style>
<div class="css-1s775nx">
</div>

`;

exports[`ssr only render a style once with styled 1`] = `

<style data-emotion-css="bjcoli">
  .css-bjcoli{color:green;}
</style>
<div class="css-bjcoli">
  <style data-emotion-css="181qz4b">
    .css-181qz4b{color:hotpink;}
  </style>
  <div class="css-181qz4b">
  </div>
  <div class="css-181qz4b">
  </div>
</div>

`;

exports[`ssr only render a style once with the css prop 1`] = `

<style data-emotion-css="bjcoli">
  .css-bjcoli{color:green;}
</style>
<div class="css-bjcoli">
  <style data-emotion-css="1lrxbo5">
    .css-1lrxbo5{color:hotpink;}
  </style>
  <div class="css-1lrxbo5">
  </div>
  <div class="css-1lrxbo5">
  </div>
</div>

`;
